pkg/controller/instance/instance_controller.go

```diff
@@ -329,7 +360,46 @@ func getParameters(instance *kudov1beta1.Instance, operatorVersion *kudov1beta1.
 	return params
 }
 
-func parameterDifference(old, new map[string]string) map[string]string {
+// makePipes generates {{ Pipes.* }} map of keys and values which is later used during template rendering.
+func makePipes(planName string, plan *v1beta1.Plan, tasks []v1beta1.Task, emeta *engine.Metadata) (map[string]string, error) {
+	taskByName := func(name string) (*v1beta1.Task, bool) {
+		for _, t := range tasks {
+			if t.Name == name {
+				return &t, true
+			}
+		}
+		return nil, false
+	}
+
+	pipes := make(map[string]string)
+
+	for _, ph := range plan.Phases {
+		for _, st := range ph.Steps {
+			for _, tn := range st.Tasks {
+				rmeta := renderer.Metadata{
+					Metadata:  *emeta,
+					PlanName:  planName,
+					PhaseName: ph.Name,
+					StepName:  st.Name,
+					TaskName:  tn,
+				}
+
+				if t, ok := taskByName(tn); ok && t.Kind == task.PipeTaskKind {
+					for _, pipe := range t.Spec.PipeTaskSpec.Pipe {
+						if _, ok := pipes[pipe.Key]; ok {
+							return nil, fmt.Errorf("duplicated pipe key %s", pipe.Key)
```

Comment:
this should totally end up as an error that is propagated into status as well as events, I don't think that's the case right now

Reply:
It totally is! See [L260](https://github.com/kudobuilder/kudo/pull/1105/files/ef7ce240ffc44c230587ea09b99521370f34786f#diff-45e4638e2c718ccd5d08635baec770b1R260)